engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala

```diff
@@ -985,41 +983,126 @@ class Compiler(
+    * Also prints the affending lines from the source along with line number - the same way as
+    * GCC does.
+    * @param diagnostic the diagnostic to pretty print
+    * @param source     the original source code
+    */
+  private class DiagnosticFormatter(
+    private val diagnostic: IR.Diagnostic,
+    private val source: Source
+  ) {
+    private val maxLineNum                     = 99999
+    private val blankLinePrefix                = "      | "
+    private val linePrefixSize                 = blankLinePrefix.length
+    private val outSupportsAnsiColors: Boolean = System.console() != null
+    private val (textAttrs: fansi.Attrs, subject: String) = diagnostic match {
+      case _: IR.Error   => (fansi.Color.Red ++ fansi.Bold.On, "error: ")
+      case _: IR.Warning => (fansi.Color.Yellow ++ fansi.Bold.On, "warning: ")
+      case _             => throw new IllegalStateException("Unexpected diagnostic type")
+    }
+    private val sourceSection: Option[SourceSection] =
+      diagnostic.location match {
+        case Some(location) =>
+          Some(source.createSection(location.start, location.length))
+        case None => None
       }
+    private val shouldPrintLineNumber = sourceSection match {
+      case Some(section) =>
+        section.getStartLine <= maxLineNum && section.getEndLine <= maxLineNum
+      case None => false
     }
 
-    if (errors.nonEmpty) {
-      output.println("Compiler encountered errors:")
-      errors.foreach { error =>
-        output.println(formatDiagnostic(error, source))
+    def format(): String = {
+      sourceSection match {
+        case Some(section) =>
+          val isOneLine = section.getStartLine == section.getEndLine
+          if (isOneLine) {
+            val lineNumber  = section.getStartLine
+            val startColumn = section.getStartColumn
+            val endColumn   = section.getEndColumn
+            var str         = fansi.Str()
+            str ++= fansi
+              .Str(source.getPath + ":" + lineNumber + ":" + startColumn + ": ")
+              .overlay(fansi.Bold.On)
+            str ++= fansi.Str(subject).overlay(textAttrs)
+            str ++= diagnostic.formattedMessage
+            str ++= "\n"
+            str ++= oneLineFromSourceColored(lineNumber, startColumn, endColumn)
+            str ++= "\n"
+            str ++= underline(startColumn, endColumn)
+            str ++= "\n"
+            str.render
+          } else {
+            var str = fansi.Str()
+            str ++= fansi
+              .Str(
+                source.getPath + ":" + section.getStartLine + ":" + section.getStartColumn + ": "
+              )
+              .overlay(fansi.Bold.On)
+            str ++= fansi.Str(subject).overlay(textAttrs)
+            str ++= diagnostic.formattedMessage
+            str ++= "\n"
+            for (lineNum <- section.getStartLine to section.getEndLine) {
+              str ++= oneLineFromSource(lineNum)
+              str ++= "\n"
+            }
```

Comment:
Do I read correctly that this will print **all** lines from a section?

If I understand correctly, a section may have hundreds of lines (if it is a definition of a type). Printing so much for a single error seems _faaar_ too much.

I recall there were plans to limit the output, no? I think we should print 2-3 lines max and then can have some kind of ellipsis `... and N more lines` for example.

Reply:
Good idea. Implemented in d1ab3773b876bae726115d33e066438ca589ca90

The ellipsis looks like this:
![image](https://github.com/enso-org/enso/assets/14013887/05683481-8d2d-49c7-a4a9-23c1d23c0258)

Pretty readable, I guess. The maximum number of lines to be printed from the source in case of multiline diagnostics is hard-coded in a variable.